ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/metrics_correlation/MetricsCorrelation.java

```diff
@@ -0,0 +1,272 @@
+import org.opensearch.ml.common.transport.model.MLModelGetResponse;
+import org.opensearch.ml.common.transport.model.MLModelSearchAction;
+import org.opensearch.ml.common.transport.register.MLRegisterModelAction;
+import org.opensearch.ml.common.transport.register.MLRegisterModelInput;
+import org.opensearch.ml.common.transport.register.MLRegisterModelRequest;
+import org.opensearch.ml.common.transport.register.MLRegisterModelResponse;
+import org.opensearch.ml.common.transport.task.MLTaskGetAction;
+import org.opensearch.ml.common.transport.task.MLTaskGetRequest;
+import org.opensearch.ml.common.transport.task.MLTaskGetResponse;
+import org.opensearch.ml.engine.algorithms.DLModelExecute;
+import org.opensearch.ml.engine.annotation.Function;
+import org.opensearch.search.SearchHit;
+import org.opensearch.search.builder.SearchSourceBuilder;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BooleanSupplier;
+
+import static org.opensearch.index.query.QueryBuilders.termQuery;
+
+@Log4j2
+@NoArgsConstructor
+@Function(FunctionName.METRICS_CORRELATION)
+public class MetricsCorrelation extends DLModelExecute {
+
+    private static final int AWAIT_BUSY_THRESHOLD = 1000;
+    private Client client;
+    private Settings settings;
+
+    private static final String MODEL_NAME = "METRICS_CORRELATION";
+    public static final String MCORR_ML_VERSION = "1.0.0";
+
+    public static final String MCORR_MODEL_URL =
+            "https://github.com/dhrubo-os/semantic-os/raw/main/mcorr.zip";
+
+    public MetricsCorrelation(Client client, Settings settings) {
+        this.client = client;
+        this.settings = settings;
+    }
+
+    @Override
+    public MetricsCorrelationOutput execute(Input input) throws Exception {
+        if (!(input instanceof MetricsCorrelationInput)) {
+            throw new IllegalArgumentException("wrong input");
+        }
+        List<MCorrModelTensors> tensorOutputs = new ArrayList<>();
+        MetricsCorrelationInput metricsCorrelation = (MetricsCorrelationInput) input;
+        List<float[]> inputData = metricsCorrelation.getInputData();
+
+        float[][] processedInputData = processedInput(inputData);
+
+        AtomicReference<String> loadedModelId = new AtomicReference<>();
+
+        SearchRequest modelSearchRequest = getSearchRequest();
+        searchModel(modelSearchRequest, ActionListener.wrap(modelInfo -> {
+            if (modelInfo.isEmpty()) {
+                registerModel(ActionListener.wrap(registerModelResponse -> {
+                    String id = getTask(registerModelResponse.getTaskId()).getModelId();
+                    loadedModelId.set(id);
+                }, e -> log.error("Final problem")));
+            } else {
+                //TODO deploy model done with state: DEPLOY_FAILED all the time but I'm getting the output
```

Comment:
Can you try on 2.x branch, not sure if this issue only happens on main branch (3.0)